README.rst
Djangorecipe: easy install of Django with buildout
==================================================

With djangorecipe you can manage your django site in a way that is familiar to
buildout users. For example:

- ``bin/django`` to run django instead of ``bin/python manage.py``.

- ``bin/test`` to run tests instead of ``bin/python manage.py test yourproject``.

- ``bin/django`` automatically uses the right django settings. So you can have
  a ``development.cfg`` buildout config and a ``production.cfg``, each telling
  djangorecipe to use a different django settings module. ``bin/django`` will
  use the right setting automatically, no need to set an environment variable.

Djangorecipe is developed on github at
https://github.com/rvanlaar/djangorecipe, you can submit bug reports there. It
is tested with travis-ci and the code quality is checked via landscape.io:


.. image:: https://secure.travis-ci.org/rvanlaar/djangorecipe.png?branch=master
   :target: http://travis-ci.org/rvanlaar/djangorecipe/

.. image:: https://landscape.io/github/rvanlaar/djangorecipe/master/landscape.svg?style=flat
   :target: https://landscape.io/github/rvanlaar/djangorecipe/master
   :alt: Code Health



Setup
-----------

You can see an example of how to use the recipe below with some of the most
common settings::

    [buildout]
    show-picked-versions = true
    parts =
        django
    eggs =
        yourproject
        gunicorn
    develop = .
    # ^^^ Assumption: the current directory is where you develop 'yourproject'.
    versions = versions

    [versions]
    Django = 1.8.2
    gunicorn = 19.3.0

    [django]
    recipe = djangorecipe
    settings = development
    eggs = ${buildout:eggs}
    project = yourproject
    test = yourproject
    scripts-with-settings = gunicorn
    # ^^^ This line generates a bin/gunicorn-with-settings script with
    # the correct django environment settings variable already set.


Earlier versions of djangorecipe used to create a project structure for you,
if you wanted it to. Django itself generates good project structures now. Just
run ``bin/django startproject <projectname>``. The main directory created is
the one where you should place your buildout and probably a ``setup.py``.

Startproject creates a ``manage.py`` script for you. You can remove it, as the
``bin/django`` script that djangorecipe creates is the (almost exact)
replacement for it.

See django's documentation for `startproject
<https://docs.djangoproject.com/en/1.8/ref/django-admin/#django-admin-startproject>`_.

You can also look at `cookiecutter <https://cookiecutter.readthedocs.org/>`_.



Supported options
-----------------

The recipe supports the following options.

project
  This option sets the name for your project.

settings
  You can set the name of the settings file which is to be used with
  this option. This is useful if you want to have a different
  production setup from your development setup. It defaults to
  `development`.

test
  If you want a script in the bin folder to run all the tests for a
  specific set of apps this is the option you would use. Set this to
  the list of app labels which you want to be tested. Normally, it is
  recommended that you use this option and set it to your project's name.

scripts-with-settings
  Script names you add to here (like 'gunicorn') get a duplicate script
  created with '-with-settings' after it (so:
  ``bin/gunicorn-with-settings``). They get the settings environment variable
  set. At the moment, it is mostly useful for gunicorn, which cannot be run
  from within the django process anymore. So the script must already be passed
  the correct settings environment variable.

  **Note**: the package the script is in must be in the "eggs" option of your
  part. So if you use gunicorn, add it there (or add it as a dependency of
  your project).

eggs
  Like most buildout recipes, you can/must pass the eggs (=python packages)
  you want to be available here. Often you'll have a list in the
  ``[buildout]`` part and re-use it here by saying ``${buildout:eggs}``.

The options below are for older projects or special cases mostly:

dotted-settings-path
  Use this option to specify a custom settings path to be used. By default,
  the ``project`` and ``settings`` option values are concatenated, so for
  instance ``myproject.development``. ``dotted-settings-path =
  somewhere.else.production`` allows you to customize it.

extra-paths
  All paths specified here will be used to extend the default Python
  path for the `bin/*` scripts. Use this if you have code somewhere without a
  proper ``setup.py``.

control-script
  The name of the script created in the bin folder. This script is the
  equivalent of the `manage.py` Django normally creates. By default it
  uses the name of the section (the part between the `[ ]`). Traditionally,
  the part is called ``[django]``.

initialization
  Specify some Python initialization code to be inserted into the
  `control-script`. This functionality is very limited. In particular, be
  aware that leading whitespace is stripped from the code given.

wsgi
  An extra script is generated in the bin folder when this is set to
  `true`. This is mostly only useful when deploying with apache's
  mod_wsgi. The name of the script is the same as the control script, but with
  ``.wsgi`` appended. So often it will be ``bin/django.wsgi``.

wsgi-script
  Use this option if you need to overwrite the name of the script above.

deploy_script_extra
  In the `wsgi` deployment script, you sometimes need to wrap the application
  in a custom wrapper for some cloud providers. This setting allows extra
  content to be appended to the end of the wsgi script. For instance
  ``application = some_extra_wrapper(application)``. The limits described
  above for `initialization` also apply here.

testrunner
  This is the name of the testrunner which will be created. It
  defaults to `test`.


Example configuration for mod_wsgi
---------------------------------------------------

If you want to deploy a project using mod_wsgi you could use this
example as a starting point::

    <Directory /path/to/buildout>
         Order deny,allow
         Allow from all
    </Directory>
    <VirtualHost 1.2.3.4:80>
         ServerName      my.rocking.server
         CustomLog       /var/log/apache2/my.rocking.server/access.log combined
         ErrorLog        /var/log/apache2/my.rocking.server/error.log
         WSGIScriptAlias / /path/to/buildout/bin/django.wsgi
    </VirtualHost>

Corner case: there is a problem when several wsgi scripts are combined in a
single virtual host instance of Apache. This is due to the fact that Django
uses the environment variable DJANGO_SETTINGS_MODULE. This variable gets set
once when the first wsgi script loads. The rest of the wsgi scripts will fail,
because they need a different settings modules. However the environment
variable DJANGO_SETTINGS_MODULE is only set once. The new `initialization`
option that has been added to djangorecipe can be used to remedy this problem
as shown below::

    [django]
    settings = acceptance
    initialization =
        import os
        os.environ['DJANGO_SETTINGS_MODULE'] = '${django:project}.${django:settings}'


Generating a control script for PyDev
---------------------------------------------------

Running Django with auto-reload in PyDev requires adding a small snippet
of code::

  import pydevd
  pydevd.patch_django_autoreload(patch_remote_debugger=False, patch_show_console=True)

just before the `if __name__ == "__main__":` in the `manage.py` module (or in
this case the control script, normally ``bin/django``, that is generated). The
following example buildout generates two control scripts: one for command-line
usage and one for PyDev, with the required snippet, using the recipe's
`initialization` option::

    [buildout]
    parts = django pydev
    eggs =
        mock

    [django]
    recipe = djangorecipe
    eggs = ${buildout:eggs}
    project = dummyshop

    [pydev]
    <= django
    initialization =
        import pydevd
        pydevd.patch_django_autoreload(patch_remote_debugger=False, patch_show_console=True)


Example usage of django-configurations
--------------------------------------

django-configurations (http://django-configurations.readthedocs.org/en/latest/)
is an application that helps you organize your Django settings into classes.
Using it requires modifying the manage.py file.  This is done easily using the
recipe's `initialization` option::

    [buildout]
    parts = django
    eggs =
        hashlib

    [django]
    recipe = djangorecipe
    eggs = ${buildout:eggs}
    project = myproject
    initialization =
        # Patch the manage file for django-configurations
        import os
        os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'myproject.settings')
        os.environ.setdefault('DJANGO_CONFIGURATION', 'Development')
        from configurations.management import execute_from_command_line
        import django
        django.core.management.execute_from_command_line = execute_from_command_line


CHANGES.rst
Changes
=======


2.3 (unreleased)
----------------

- Nothing changed yet.


2.2.1 (2016-06-29)
------------------

- Bugfix for 2.2: ``bin/test`` was missing quotes around an option. [reinout]


2.2 (2016-06-29)
----------------

- Added optional ``coverage`` option. Set it to ``true`` to automatically run
  coverage around your django tests. Needed if you used to have a test runner
  like django-nose run your coverage automatically. Since django 1.7, this
  doesn't work anymore. With the new "coverage" option, ``bin/test`` does it
  for you. [reinout]

- Automated tests (travis-ci.org) test with django 1.4, 1.8 and 1.9 now. And
  pypi, python 2.7 and python 3.4. [reinout]

2.1.2 (2015-10-21)
------------------

- Fixed documentation bug: the readme mentioned ``script-with-settings``
  instead of ``scripts-with-settings`` (note the missing ``s`` after
  ``script``). The correct one is ``script-with-settings``.
  [tzicatl]


2.1.1 (2015-06-15)
------------------

- Bugfix: ``script-entrypoints`` entry point finding now actually works.


2.1 (2015-06-15)
----------------

- Renamed ``script-entrypoints`` option to ``scripts-with-settings``. It
  accepts script names that would otherwise get generated (like ``gunicorn``)
  and generates a duplicate script named like ``bin/gunicorn-with-settings``.

  Technical note: this depends on the scripts being setuptools "console_script
  entrypoint" scripts.


2.0 (2015-06-10)
----------------

- Removed project generation. Previously, djangorecipe would generate a
  directory for you from a template, but Django's own template is more than
  good enough now. Especially: it generates a subdirectory for your project
  now. Just run ``bin/django startproject <projectname>``.

  See django's documentation for `startproject
  <https://docs.djangoproject.com/en/1.8/ref/django-admin/#django-admin-startproject>`_.

  You can also look at `cookiecutter <https://cookiecutter.readthedocs.org/>`_.

  This also means the ``projectegg`` option is now deprecated, it isn't needed
  anymore.

- We aim at django 1.7 and 1.8 now. Django 1.4 still works, (except that that
  one doesn't have a good startproject command).

- Gunicorn doesn't come with the django manage.py integration, so ``bin/django
  run_gunicorn`` doesn't work anymore. If you add ``script-entrypoints =
  gunicorn`` to the configuration, we generate a ``bin/django_env_gunicorn``
  script that is identical to ``bin/gunicorn``, only with the environment
  correctly set.  **Note: renamed in 2.1 to ``scripts-with-settings``**.

  This way, you can use the wsgi.py script in your project (copy it from the
  django docs if needed) with ``bin/django_env_gunicorn yourproject/wsgi.py``
  just like suggested everywhere. This way you can adjust your wsgi file to
  your liking and run it with gunicorn.

  For other wsgi runners (or programs you want to use with the correct
  environment set), you can add a full entry point to ``script-entrypoints``,
  like ``script-entrypoints = gunicorn=gunicorn.app.wsgiapp:run`` would be the
  full line for gunicorn. Look up the correct entrypoint in the relevant
  package's ``setup.py``.

  Django's 1.8 ``wsgi.py`` file looks like this, see https://docs.djangoproject.com/en/1.8/howto/deployment/wsgi/::

      import os

      from django.core.wsgi import get_wsgi_application

      os.environ.setdefault("DJANGO_SETTINGS_MODULE", "yourproject.settings")

      application = get_wsgi_application()


- The ``wsgilog`` option has been deprecated, the old apache mod_wsgi script
  hasn't been used for a long time.

- Removed old pth option, previously used for pinax. Pinax uses proper python
  packages since a long time, so it isn't needed anymore.



1.11 (2014-11-21)
-----------------

- The ``dotted-settings-path`` options was only used in management script. Now
  it is also used for the generated wsgi file and the test scripts.


1.10 (2014-06-16)
-----------------

- Added ``dotted-settings-path`` option. Useful when you want to specify a
  custom settings path to be used by the ``manage.main()`` command.

- Renamed ``deploy_script_extra`` (with underscores) to
  ``deploy-script-extra`` (with dashes) for consistency with the other
  options. If the underscore version is found, an exception is raised.


1.9 (2014-05-27)
----------------

- ``bin/test`` now passes along command line arguments to the underlying
  management command. Previously, only the equivalent of ``manage.py test
  list_of_apps`` would be done. Now command line arguments are passed as-is
  after the list of apps.

- Added ``deploy_script_extra`` option. It is appended to the wsgi script.
  Useful for instance for a cloud hoster where you need to wrap your wsgi
  application object in a custom call.


1.8 (2014-05-27)
----------------

- Supporting buildout's relative-paths setting now.


1.7 (2013-12-11)
----------------

- Add option to change the wsgi script name. Thanks hedleyroos (Fixes pull #74)

1.6 (2013-10-28)
----------------

- Djangorecipe works with django 1.6 now.

- Tested with Django 1.4, 1.5 and 1.6. Pre-1.4 support is gone, now. Also
  tested on Python 2.6/2.7, 3.2/3.3.

- Moved to buildout 2 support only.

- Removed generation of fastcgi script. You can run it easily as ``bin/django
  runfcgi`` and it will be deprecated in Django 1.7 anyway.


1.5 (2013-01-25)
----------------

- Removed support for a different python version than the one you use to run
  buildout with. Previously, you could run your buildout with 2.6 but get
  Django to use 2.7 instead. zc.buildout 2.0 doesn't allow it anymore, so we
  removed it too.


1.4 (2013-01-15)
----------------

- Added initialization code support. Thanks to anshumanb, jjmurre. (Closes #58).


1.3 (2012-09-07)
----------------

- Removed deprecation warning in Django 1.4. Fixes #49, thanks Shagi.

- Added documentation for use with mr.developer. Thanks shagi (closes issue #45)

- Added Travis support.


1.2.1 (2012-05-15)
------------------


src/djangorecipe/recipe.py
import os
import logging
import sys

from zc.buildout import UserError
import pkg_resources
import zc.recipe.egg

from djangorecipe.boilerplate import WSGI_TEMPLATE


class Recipe(object):
    def __init__(self, buildout, name, options):
        self.log = logging.getLogger(name)

        # Deprecations
        if 'version' in options:
            raise UserError('The version option is deprecated. '
                            'Read about the change on '
                            'http://pypi.python.org/pypi/djangorecipe/0.99')
        if 'wsgilog' in options:
            raise UserError('The wsgilog option is deprecated. '
                            'Read about the change on '
                            'http://pypi.python.org/pypi/djangorecipe/2.0')
        if 'projectegg' in options:
            raise UserError("The projectegg option is deprecated. "
                            "See the changelog for 2.0 at "
                            "http://pypi.python.org/pypi/djangorecipe/2.0")
        if 'deploy_script_extra' in options:
            # Renamed between 1.9 and 1.10
            raise UserError(
                "'deploy_script_extra' option found (with underscores). "
                "This has been renamed to 'deploy-script-extra'.")
        if 'script-entrypoints' in options:
            # Renamed between 2.0 and 2.1
            raise UserError("The 'script-entrypoints' option is deprecated by "
                            "'scripts-with-settings'. See the changelog "
                            "for 2.1 at "
                            "http://pypi.python.org/pypi/djangorecipe/2.1")

        # Generic initialization.
        self.egg = zc.recipe.egg.Egg(buildout, options['recipe'], options)
        self.buildout, self.name, self.options = buildout, name, options
        options['location'] = os.path.join(
            buildout['buildout']['parts-directory'], name)
        options['bin-directory'] = buildout['buildout']['bin-directory']

        # Option defaults.
        options.setdefault('project', 'project')
        options.setdefault('settings', 'development')
        options.setdefault('extra-paths', '')
        options.setdefault('initialization', '')
        options.setdefault('deploy-script-extra', '')
        options.setdefault('scripts-with-settings', '')
        options.setdefault('coverage', '')

        # mod_wsgi support script
        options.setdefault('wsgi', 'false')
        options.setdefault('logfile', '')

        # respect relative-paths (from zc.recipe.egg)
        relative_paths = options.get(
            'relative-paths', buildout['buildout'].get('relative-paths',
                                                       'false'))
        if relative_paths == 'true':
            options['buildout-directory'] = buildout['buildout']['directory']
            self._relative_paths = options['buildout-directory']
        else:
            self._relative_paths = ''
            assert relative_paths == 'false'

    def install(self):
        if self.options['project'] not in os.listdir(
                self.buildout['buildout']['directory']):
            # Only warn for this upon install, not on update.
            self.log.warn(
                "There's no directory named after our project. "
                "Probably you want to run 'bin/django startproject %s'",
                self.options['project'])

        extra_paths = self.get_extra_paths()
        ws = self.egg.working_set(['djangorecipe'])[1]
        # ^^^ working_set returns (requirements, ws)

        script_paths = []
        script_paths.extend(self.create_manage_script(extra_paths, ws))
        script_paths.extend(self.create_test_runner(extra_paths, ws))
        script_paths.extend(self.make_wsgi_script(extra_paths, ws))
        script_paths += self.create_scripts_with_settings(
            extra_paths, ws)
        return script_paths

    def create_manage_script(self, extra_paths, ws):
        settings = self.get_settings()
        return zc.buildout.easy_install.scripts(
            [(self.options.get('control-script', self.name),
              'djangorecipe.binscripts', 'manage')],
            ws, sys.executable, self.options['bin-directory'],
            extra_paths=extra_paths,
            relative_paths=self._relative_paths,
            arguments="'%s'" % settings,
            initialization=self.options['initialization'])

    def create_test_runner(self, extra_paths, working_set):
        settings = self.get_settings()
        coverage_functions = self.options.get('coverage', '')
        if coverage_functions.lower() == 'true':
            coverage_functions = 'report html_report xml_report'
        apps = self.options.get('test', '').split()
        # Only create the testrunner if the user requests it
        if apps:
            return zc.buildout.easy_install.scripts(
                [(self.options.get('testrunner', 'test'),
                  'djangorecipe.binscripts', 'test')],
                working_set, sys.executable,
                self.options['bin-directory'],
                extra_paths=extra_paths,
                relative_paths=self._relative_paths,
                arguments="'%s', '%s', %s" % (
                    settings,
                    coverage_functions,
                    ', '.join(["'%s'" % app for app in apps])),
                initialization=self.options['initialization'])
        else:
            return []

    def make_wsgi_script(self, extra_paths, ws):
        scripts = []
        _script_template = zc.buildout.easy_install.script_template
        settings = self.get_settings()
        zc.buildout.easy_install.script_template = (
            zc.buildout.easy_install.script_header +
            WSGI_TEMPLATE +
            self.options['deploy-script-extra']
        )
        if self.options.get('wsgi', '').lower() == 'true':
            scripts.extend(
                zc.buildout.easy_install.scripts(
                    [(self.options.get('wsgi-script') or
                      '%s.%s' % (self.options.get('control-script',
                                                  self.name),
                                 'wsgi'),
                      'djangorecipe.binscripts', 'wsgi')],
                    ws,
                    sys.executable,
                    self.options['bin-directory'],
                    extra_paths=extra_paths,
                    relative_paths=self._relative_paths,
                    arguments="'%s', logfile='%s'" % (
                        settings, self.options.get('logfile')),
                    initialization=self.options['initialization'],
                ))
        zc.buildout.easy_install.script_template = _script_template
        return scripts

    def create_scripts_with_settings(self, extra_paths, ws):
        """Create duplicates of existing scripts - *with* a settings env.

        What we're installing here are existing setuptools entry points. We
        look up the script names in the list of available entry points and
        install a duplicate. We postfix the duplicate with '-with-settings',
        so that 'gunicorn' for instance becomes 'gunicorn-with-settings'.

        """
        script_names = [entrypoint.strip() for entrypoint in
                        self.options.get('scripts-with-settings').splitlines()
                        if entrypoint.strip()]
        if not script_names:
            return []
        settings = self.get_settings()
        postfix = '-with-settings'
        initialization = self.options['initialization']
        initialization += (
            "\n" +
            "import os\n" +
            "os.environ['DJANGO_SETTINGS_MODULE'] = '%s'" % settings)
        created_scripts = []
        known_entrypoints = list(ws.iter_entry_points('console_scripts'))
        to_create = [entrypoint for entrypoint in known_entrypoints
                     if entrypoint.name in script_names]
        for entrypoint in to_create:
            script_name = entrypoint.name + postfix
            dotted_path = entrypoint.module_name
            function_name = entrypoint.attrs[0]
            self.log.debug("Creating entrypoint %s:%s as %s",
                           dotted_path, function_name, script_name)
            zc.buildout.easy_install.scripts(
                [(script_name, dotted_path, function_name)],
                ws, sys.executable, self.options['bin-directory'],
                extra_paths=extra_paths,
                relative_paths=self._relative_paths,
                initialization=initialization)
            created_scripts.append(script_name)

        # Feedback logging.
        known_names = [entrypoint.name for entrypoint in known_entrypoints]
        unkown_script_names = [name for name in script_names
                               if name not in known_names]
        if unkown_script_names:
            raise UserError("Some script names couldn't be found: %s" % (
                ', '.join(unkown_script_names)))

        return created_scripts

    def get_extra_paths(self):
        extra_paths = [self.buildout['buildout']['directory']]
        pythonpath = [p.replace('/', os.path.sep) for p in
                      self.options['extra-paths'].splitlines() if p.strip()]
        extra_paths.extend(pythonpath)
        return extra_paths

    def update(self):
        extra_paths = self.get_extra_paths()
        ws = self.egg.working_set(['djangorecipe'])[1]
        # ^^^ working_set returns (requirements, ws)

        self.create_manage_script(extra_paths, ws)
        self.create_test_runner(extra_paths, ws)
        self.make_wsgi_script(extra_paths, ws)

    def create_file(self, filename, template, options):
        if os.path.exists(filename):
            return
        f = open(filename, 'w')
        f.write(template % options)
        f.close()

    def get_settings(self):
        settings = '%s.%s' % (self.options['project'], self.options['settings'])
        settings = self.options.get('dotted-settings-path', settings)
        return settings
